Add DTLS handshake fragmentation test

piotr-now opened the pull request:
## Description
Add test for DTLS handshake fragmentation

## Status
**READY**

Patater committed:
pkey: Remove dependency on X.509

Commit:
Deprecate MBEDTLS_SSL_HW_RECORD_ACCEL config

Signed-off-by: Manuel Pégourié-Gonnard <manuel.pegourie-gonnard@arm.com>

Commit:
Deprecate global func ptrs for MBEDTLS_SSL_HW_RECORD_ACCEL

Signed-off-by: Manuel Pégourié-Gonnard <manuel.pegourie-gonnard@arm.com>

Commit:
Wrap deprecations for HW_RECORD_ACCEL with DEPRECATED_REMOVED

Signed-off-by: Manuel Pégourié-Gonnard <manuel.pegourie-gonnard@arm.com>

Commit:
Remove the word likely from deprecated documentation

Signed-off-by: Manuel Pégourié-Gonnard <manuel.pegourie-gonnard@arm.com>

Commit:
Add comment to matching #endif

Signed-off-by: Manuel Pégourié-Gonnard <manuel.pegourie-gonnard@arm.com>

Patater committed:
pkey/rsa_genkey: Remove commented out code

There is some commented out X.509 certificate writing code present in
rsa_genkey. It looks like it has been commented out since the beginning
of time. Let's remove it, since commented out code is not in good style.

Patater committed:
query_config: Move to programs/test

As the SSL programs, like ssl_client2 and ssl_server2, are dependent on
SSL and therefore about to be removed, the only consumer of query_config
is the query_compile_time_config test. As such, it makes sense to move
query_config to be next to what uses it.

apple-ihack-geek committed:
Fixed possibly undefined variable warnings by initializing variables to 0.

gilles-peskine-arm committed:
Add a calloc self-test

Add a very basic test of calloc to the selftest program. The selftest
program acts in its capacity as a platform compatibility checker rather
than in its capacity as a test of the library.

The main objective is to report whether calloc returns NULL for a size
of 0. Also observe whether a free/alloc sequence returns the address
that was just freed and whether a size overflow is properly detected.

gilles-peskine-arm committed:
Add a test component with malloc(0) returning NULL

Exercise the library functions with calloc returning NULL for a size
of 0. Make this a separate job with UBSan (and ASan) to detect
places where we try to dereference the result of calloc(0) or to do
things like

    buf = calloc(size, 1);
    if (buf == NULL && size != 0) return INSUFFICIENT_MEMORY;
    memcpy(buf, source, size);

which has undefined behavior when buf is NULL at the memcpy call even
if size is 0.

This is needed because other test components jobs either use the system
malloc which returns non-NULL on Linux and FreeBSD, or the
memory_buffer_alloc malloc which returns NULL but does not give as
useful feedback with ASan (because the whole heap is a single C
object).

gilles-peskine-arm committed:
Fix uninitialized variable in an edge case

If `context_buf = mbedtls_calloc( 1, buf_len )` failed,
`context_buf_len` was not initialized. Noticed by
`gcc -Os -Werror=maybe-uninitialized`.

This was only a problem in ssl_server2 (a test program), only with
MBEDTLS_SSL_CONTEXT_SERIALIZATION enabled.

gilles-peskine-arm committed:
Replace -O0 by -O1 or -Os in most components

Gcc skips some analyses when compiling with -O0, so we may miss
warnings about things like uninitialized variables.

gilles-peskine-arm committed:
Test GCC and Clang with common build options

Goals:
* Build with common compilers with common options, so that we don't
  miss a (potentially useful) warning only triggered with certain
  build options.
* A previous commit removed -O0 test jobs, leaving only the one with
  -m32. We have inline assembly that is disabled with -O0, falling
  back to generic C code. This commit restores a test that runs the
  generic C code on a 64-bit platform.

gilles-peskine-arm committed:
Add ChangeLog entry for #2663

gilles-peskine-arm committed:
Fix CTR_DRBG benchmark

You can't reuse a CTR_DRBG context without free()ing it and
re-init()ing. This generally happened to work, but was never
guaranteed. It could have failed with alternative implementations of
the AES module because mbedtls_ctr_drbg_seed() calls
mbedtls_aes_init() on a context which is already initialized if
mbedtls_ctr_drbg_seed() hasn't been called before, plausibly causing a
memory leak. Calling free() and seed() with no intervening init fails
when MBEDTLS_THREADING_C is enabled and all-bits-zero is not a valid
mutex representation. So add the missing free() and init().

Commit:
Move ssl_set_timer() to public namespace

Commit:
Move ssl_ep_len() to public namespace

Commit:
Move ssl_check_timer() to public namespace

Commit:
Move SSL internal pointer updating routines to public namespace

Commit:
Move ssl_session_reset_int() to public namespace

Commit:
Move ssl_dtls_replay_reset() to public namespace

Commit:
Move ssl_resend_hello_request() to public namespace

Commit:
Move ssl_handshake_wrapup_free_hs_transform() to public namespace

Commit:
Move ssl_start_renegotiation() to public namespace

Commit:
Move flight and message buffering functions to public namespace

Commit:
Move ssl_get_current_mtu() to public namespace

Commit:
Duplicate ssl_tls.c as ssl_tls_old.c and ssl_msg.c

This commit is the first in a series of commits aiming to split
the content of ssl_tls.c in two files ssl_tls.c and ssl_msg.c.

As a first step, this commit replaces ssl_tls.c by two identical
copies ssl_tls_old.c and ssl_msg.c. Even though the file
ssl_tls_old.c will subsequently be renamed back into ssl_tls.c,
this approach retains the git history in both files.

Commit:
Rename temporary copy ssl_tls_old.c back to ssl_tls.c

Commit:
Add guarded ssl_msg.c to Makefiles

This commit adds the newly created copy ssl_msg.c of ssl_tls.c
to the build system but guards its content by an `#if 0 ... #endif`
preprocessor guard in order to avoid compilation failures resulting
from code duplication. This guard will be removed once the contents
of ssl_tls.c and ssl_msg.c have been made disjoint.

Commit:
Remove code from ssl_tls.c & ssl_msg.c to get disjoint functionality

This commit is the final step in separating the functionality of
what was originally ssl_tls.c into both ssl_tls.c and ssl_msg.c.

So far, ssl_msg.c has been created as an identical copy of ssl_tls.c.
For each block of code in these files, this commit removes it from
precisely one of the two files, depending on where the respective
functionality belongs.

The splitting separates the following functionalities:

1) An implementation of the TLS and DTLS messaging layer, that is,
   the record layer as well as the DTLS retransmission state machine.

   This is now contained in ssl_msg.c

2) Handshake parsing and writing functions shared between client and
   server (functions specific to either client or server are implemented
   in ssl_cli.c and ssl_srv.c, respectively).

   This is remains in ssl_tls.c.

Commit:
Adapt preamble for newly created ssl_msg.c

Commit:
Move ssl_mac() from ssl_tls.c to ssl_msg.c

mpg committed:
Merge pull request #2852 from gilles-peskine-arm/2.19-fix-full-Os

Fix and test the full config with gcc and clang

gilles-peskine-arm committed:
Clarify two identical changelog entries

The contribution from #2663 was split in two: the crypto part was
mereged in 2.19.1 and the x509 part was merged after 2.20.0. Tweak the
wording of the changelog entries to specify which is which.

Commit:
Add compile-time guards around declarations in ssl_internal.h

gilles-peskine-arm committed:
cmake: link programs that only use crypto with libmbedcrypto

When building with CMake, for sample programs that only use
functionality in libmbedcrypto (i.e. crypto and platform), link with
libmbedcrypto, not with libmbedtls.

This doesn't change the result, because the linker skips libraries in
which no symbol is used, but it changes the build dependencies, and it
has the advantage of bringing programs/*/CMakeLists.txt closer to the
corresponding files under crypto/.

The programs concerned are crypto sample and test programs, and
programs that only use (potential) platform functions such as
mbedtls_printf. dh_client and dh_server keep linking with mbedtls
because they use functions from the net_sockets module.

gilles-peskine-arm committed:
cmake: link programs that only use x509 with libmbedx509

When building with CMake, for sample programs that only use
functionality in libmbedcrypto and libmbedx509, link with
libmbedx509, not with libmbedtls.

cert_app makes a TLS connection, so do link it with libmbedtls.

gilles-peskine-arm committed:
programs/Makefile: List all programs one by one

This makes it easier to add or remove programs as well as see which
programs were added or removed in diffs.

Side port of 30fae8ee7dc247116da968aa07d8164eaa6f07af in mbed-crypto.

gilles-peskine-arm committed:
Test component with malloc(0) returning NULL: run some ssl tests

gilles-peskine-arm committed:
Update crypto submodule

* #352: Parse RSA parameters DP, DQ and QP from PKCS1 private keys
* #263: Introduce ASN.1 SEQUENCE traversal API
* #345: Fix possible error code mangling in psa_mac_verify_finish
* #357: Update Mbed Crypto with latest Mbed TLS changes as of 2020-02-03
* #350: test_suite_asn1parse: improve testing of trailing garbage in parse_prefixes
* #346: Improve robustness and testing of mbedtls_mpi_copy

gilles-peskine-arm committed:
Update ChangeLog for crypto changes since Mbed Crypto 3.0.1

Add ChangeLog entries for changes brought by the submodule update in
81d31002503f0cdc735176279e913106155666c4.

gilles-peskine-arm committed:
Update ChangeLog for crypto changes from 799ae77f90584f39c2fbdae7e117ff5742774d77

Add ChangeLog entries for changes brought by the submodule update in
the previous commit.

mpg committed:
Merge pull request #2873 from gilles-peskine-arm/bk-warning-fixes-x509

Fix some possibly-undefined variable warnings

mpg committed:
Merge pull request #3010 from hanno-arm/tls_msg_split

Introduce separate source file for SSL messaging layer

piotr-now committed:
Add test with sending application data via DTLS

Signed-off-by: Piotr Nowicki <piotr.nowicki@arm.com>

Commit:
Add a dtls handshake test with context serialization

Signed-off-by: Andrzej Kurek <andrzej.kurek@arm.com>

piotr-now committed:
Add test for renegotiation in DTLS

Signed-off-by: Piotr Nowicki <piotr.nowicki@arm.com>

gilles-peskine-arm committed:
Merge pull request #3037 from gilles-peskine-arm/update-submodule-20200206

Update crypto submodule 2020-02-06

yanesca committed:
Update submodule

* #365 Change PSA compatibility API to inline functions
* #367 Fix pk_parse_key()'s use of rsa_complete()
* #370 Bump version to Mbed TLS 2.21.0

yanesca committed:
Update ChangeLog for crypto changes from a591985c62a9354db3027b4895c24377712b9620

Add ChangeLog entries for changes brought by the submodule update in
the previous commit.

yanesca committed:
Add release info to ChangeLog

yanesca committed:
Bump version to Mbed TLS 2.21.0

yanesca committed:
Merge pull request #3055 from yanesca/update_submodule_and_version_2.21.0

Update submodule and version 2.21.0

piotr-now committed:
Add DTLS handshake fragmentation test

For this test it is good to have a handshake messages length as big as
possible, so for the server the certificate verification mode is
changed from default NONE to REQUIRED. It requires the client to send
certificate date to the server during handshake

Signed-off-by: Piotr Nowicki <piotr.nowicki@arm.com>

Patater committed:
Merge pull request #3036 from AndrzejKurek/dtls-handshake-tests

DTLS handshake tests

Patater commented on tests/suites/test_suite_ssl.function:
```diff
@@ -3680,6 +3737,44 @@ void handshake_serialization( )
+        TEST_ASSERT( cli_hs_fragmented >= 1 );
+    }
+    /* The goto below is used to avoid an "unused label" warning.*/
+    goto exit;
```
This `goto` seems weird. Why do we need it here but not in other tests?

piotr-now replied:
Yes, this `goto exit;` is unnecessary here because it is already in TEST_ASSERT. In the other tests it is needed to prevent warning. After Andrzej's refactoring some of tests does not have any TEST_ASSERT, but this `exit:` label is added automatically during compilation (by script). I will remove it

piotr-now replied:
Removed

Patater commented on tests/suites/test_suite_ssl.function:
```diff
@@ -1651,6 +1697,9 @@ void perform_handshake( handshake_test_options* options )
                                             options->pk_alg, NULL, NULL, NULL ) == 0 );
     }
+
+    mbedtls_ssl_conf_authmode( &server.conf, options->srv_auth_mode );
```
Why was this added? What does it have to do with handshake fragmentation testing? Should this be added in its own commit?

piotr-now replied:
By default, server does not check client's certificates so I had to set it manually to be required. This forces the client to send more data to the server during handshake. Without that handshake messages from the client would be to small to be fragmented.

Patater replied:
OK, I see how it is related to fragmentation now. Could you please explain this aspect of your implementation in the git commit message?

piotr-now replied:
I added an explanation to the commit message

yanesca committed:
Merge pull request #3038 from gilles-peskine-arm/align-crypto-programs

Side-port from crypto: programs/

AndrzejKurek commented on tests/suites/test_suite_ssl.function:
```diff
@@ -3680,6 +3737,42 @@ void handshake_serialization( )
+    /* Test if the server received fragmented handshake */
+    if( expected_srv_hs_fragmentation )
+    {
+        TEST_ASSERT( srv_hs_fragmented >= 1 );
```
Is this sufficient? Why won't we check for exact values, passed as a test parameter?

piotr-now replied:
I was thinking about it too, but it is pretty hard to determinate how much data should be transferred during handshake and in how many fragments. I decided that the fact of any fragmentation is sufficient enough for this test.

AndrzejKurek replied:
What do you mean by `hard to determine`? Is it time consuming for the programmer, is it non-deterministic, or is there no way to tell the exact number of fragments? I can see that in `ssl-opt.sh` application data tests check the exact number and size of fragments, while DTLS handshake fragmentation tests only check if the handshake message was fragmented, so I'm inclined towards accepting it, but I would also like to know the background a bit more.

piotr-now replied:
If we want to send application data, we decide how much of them should be. For handshake the message length is unknown it should be estimate. It is related with the certificates, allowed cipher suites and etc. I think it is not easy to estimate the biggest message size during handshake. I could hardcoded the current number of fragments with which this test is performed (if it is good enough)

AndrzejKurek commented on tests/suites/test_suite_ssl.function:
```diff
@@ -6,13 +6,42 @@
+{
+    (void)level; (void)line; (void)file;
+
+    log_pattern *p = (log_pattern*)ctx;
```
Space missing

AndrzejKurek commented on tests/suites/test_suite_ssl.function:
```diff
@@ -6,13 +6,42 @@
+        NULL != p->counter &&
+        NULL != strstr( str, p->pattern ) )
+    {
+        ++(*( p->counter ) );
```
Space missing

Patater committed:
Merge pull request #3042 from AndrzejKurek/dtls-handshake-serialization

Dtls handshake serialization test

Patater commented on tests/suites/test_suite_ssl.function:
```diff
@@ -3680,6 +3737,42 @@ void handshake_serialization( )
+
+    srv_pattern.pattern = cli_pattern.pattern = "found fragmented DTLS handshake";
+    srv_pattern.counter = &srv_hs_fragmented;
+    cli_pattern.counter = &cli_hs_fragmented;
```
Where is `cli_pattern.pattern` set?

AndrzejKurek commented on tests/suites/test_suite_ssl.function:
```diff
@@ -6,13 +6,42 @@
+} log_pattern;
+
+/*
+ * This function can be added as a debug output for log messages from mbedtls
```
This is a bit unclear, I would rather state that: 
```suggestion
 * This function can be passed to mbedtls to receive output logs from it. In this case, it will count the instances of a log_pattern in the received logged messages.
```

AndrzejKurek commented on tests/suites/test_suite_ssl.function:
```diff
@@ -3680,6 +3737,42 @@ void handshake_serialization( )
+
+    perform_handshake( &options );
+
+    /* Test if the server received fragmented handshake */
```
```suggestion
    /* Test if the server received a fragmented handshake */
```

AndrzejKurek commented on tests/suites/test_suite_ssl.function:
```diff
@@ -3680,6 +3737,42 @@ void handshake_serialization( )
+    {
+        TEST_ASSERT( srv_hs_fragmented >= 1 );
+    }
+    /* Test if the client received fragmented handshake */
```
```suggestion
    /* Test if the client received a fragmented handshake */
```

AndrzejKurek reviewed (changes requested):
Minor changes requested.

AndrzejKurek commented on tests/suites/test_suite_ssl.function:
```diff
@@ -6,13 +6,42 @@
+                          const char *file, int line,
+                          const char *str )
+{
+    (void)level; (void)line; (void)file;
```
Please move those to separate lines, and add a space after `(void)`, as I'm pretty sure I haven't seen this convention used anywhere in the library.

mpg committed:
Uniformize wording for removal of deprecated things

Signed-off-by: Manuel Pégourié-Gonnard <manuel.pegourie-gonnard@arm.com>

piotr-now commented:
I changed counter inside log_pattern to be a variable instead of pointer to variable

AndrzejKurek commented on tests/suites/test_suite_ssl.function:
```diff
@@ -5,6 +5,35 @@
+    (void) file;
+
+    log_pattern *p = (log_pattern*) ctx;
+    if( NULL != p &&
```
This formatting is a bit strange (since you're well below the line limit), but it's a minor.

Patater committed:
Merge pull request #3043 from piotr-now/dtls-application-data

DTLS application data

Patater committed:
Merge pull request #3046 from piotr-now/dtls-renegotiation

DTLS renegotiation

Patater reviewed (approved):
LGTM

mpg committed:
Merge pull request #2263 from andresag01/iotssl-2544-deprecate-record-accel-2

Deprecate feature MBEDTLS_SSL_HW_RECORD_ACCEL

Commit:
test_suite_ssl refactoring: merge tls & dtls application data tests

Move the app data tests to a shared handshake function to simplify further 
addition of tests.
Signed-off-by: Andrzej Kurek <andrzej.kurek@arm.com>

Commit:
test_suite_ssl refactoring: merge renegotiation test into handshake

Move the renegotiation test to a shared handshake function to simplify further
addition of tests.
Signed-off-by: Andrzej Kurek <andrzej.kurek@arm.com>

Commit:
test_suite_ssl refactoring: provide default options structure for tests

Create and provide a structure with default options so that the caller won't have
to pass all of the parameters each time the handshake is called. In the future
this can be improved so that the options are passed as a string, just like in
ssl-opt.sh.
Signed-off-by: Andrzej Kurek <andrzej.kurek@arm.com>
